### ISSUE TITLE
Add rule to warn about use of `arguments` in fat arrow functions

### ISSUE THREAD

**Issue by WickyNilliams** [eslint/eslint#8639]
Please describe what the rule should do:

warn on usage of arguments inside fat arrow functions

What category of rule is this? (place an "X" next to just one item)

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

Provide 2-3 code examples that this rule will warn about:

```js
const logArgs = () => console.log(arguments);
```

```js
function foo(a, b, c) {
  const bar = blah => {
    if(arguments.length > 1) {
      console.log("too many args!");
    }
  };
  bar(123);
}

foo(456, 789);
```

Why should this rule be included in ESLint (instead of a plugin)?

`arguments` is not bound for fat arrow functions. The first example is quite easy to catch by a developer, as an error will be thrown. The second example would be hard to track down because `arguments` is actually referring to `foo`'s arguments, not `bar`'s

**Comment by platinumazure** [comment]
I'm surprised we don't have something like this already. I agree this would be a great rule to add. :+1: from me.

**Comment by WickyNilliams** [comment]
Me too! I searched long and hard through the rules, absolutely convinced that there must be something for this already. Really surprised there isn't

**Comment by not-an-aardvark** [comment]
Would the rule also report using `this`, `super`, and `new.target` in arrow functions?

My concern is that a user might have intentionally used an arrow function to get this behavior. Contrived example:

```js
// Returns a function that calls `fetch` with the given arguments
function getFetchFunction() {
    return () => fetch(...arguments);
}
```

If this case is exceedingly rare compared to the case where the pattern was an error, then people can just use disable comments, but I'm not sure whether it's actually rare compared to the error case.

**Comment by platinumazure** [comment]
@not-an-aardvark Even in that case, it's unclear for future maintainers.

It would be easy to work around and make the code clearer:

```js
// Returns a function that calls `fetch` with the given arguments to this function
function getFetchFunction(...args) {
    return () => fetch(...args);
}
```

**Comment by WickyNilliams** [comment]
or even in ES5:

```js
function getFetchFunction() {
    var args = arguments;
    return () => fetch(args);
}
```

Both of these are much clearer

**Comment by not-an-aardvark** [comment]
Hmm, I'm not convinced.

> I would expect lots of people use `this` in arrow functions and know what it means, but maybe not as much for some of the other types.

Sure, but I think that's just because `this` is used more often than `arguments`. I don't see why using `arguments` in arrow functions is worse than using `this` in arrow functions.

> It would be easy to work around and make the code clearer:
> ```js
>// Returns a function that calls `fetch` with the given arguments to this function
>function getFetchFunction(...args) {
>    return () => fetch(...args);
>}

If people can use rest parameters then I don't think they should use `arguments` at all. That's why we have [`prefer-rest-params`](http://eslint.org/docs/rules/prefer-rest-params).

> or even in ES5:
>```js
>function getFetchFunction() {
>    var args = arguments;
>    return () => fetch.apply(null, args);
>}
>```

Arrow functions are invalid syntax in ES5.

**Comment by WickyNilliams** [comment]
> Arrow functions are invalid syntax in ES5.

D'oh!

My point with this rule suggestion was that:

a) such code is more likely an error than intentional (if it is intentional, see (c))
b) it's confusing/difficult to track down because `arguments` might refer to something *n* levels up the stack
c) every instance can be re-written to be more clear via spread or other methods

**Comment by aboyton** [comment]
There is the rule `prefer-rest-params` which warns about use of `arguments` anywhere (not just fat arrow functions). I'm curious why you wouldn't just use that rule rather than making a specific one for this use case.

**Comment by WickyNilliams** [comment]
Now I've thought of it more, I think you're right. May as well use rest params in every case. Closing
